lib/bridge/IrcBridge.js

```diff
@@ -754,6 +769,128 @@ IrcBridge.prototype.getBotClient = function(server) {
     });
 }
 
+
+var _solveTimeout = null;
+
+IrcBridge.prototype._updateVisibilityMap = function(isMode, key, value) {
+    let hasChanged = false;
+    if (isMode) {
+        if (typeof value !== 'boolean') {
+            throw new Error('+s state must be indicated with a boolean');
+        }
+        if (this._visibilityMap.channelModes[key] !== value) {
+            this._visibilityMap.channelModes[key] = value;
+            hasChanged = true;
+        }
+    }
+    else {
+        if (typeof value !== 'string' || (value !== "private" && value !== "public")) {
+            throw new Error('Room visibility must = "private" | "public"');
+        }
+
+        if (this._visibilityMap.roomVisibilities[key] !== value) {
+            this._visibilityMap.roomVisibilities[key] = value;
+            hasChanged = true;
+        }
+    }
+
+    if (hasChanged) {
+        clearTimeout(_solveTimeout);
+
+        _solveTimeout = setTimeout(() => {
+            this._solveVisibility()
+        }, 10000);
+        return Promise.resolve();
+    }
+
+    return Promise.resolve();
+}
+
+IrcBridge.prototype._solveVisibility = Promise.coroutine(function*() {
+    // For each room, do a big OR on all of the channels that are linked in any way
```

Comment:
Obviously we both know what we mean by "solve visibility" but the rest of the team won't. I would go in to a decent amount of detail explaining what the problem is, why you need to solve it, and how this function solves it. As a very terse example:
```js
// Walk the graph of mapped rooms to work out if a Matrix room should be on the published room list or not.
//
// We need to do this in order to ensure that messages sent on a +s (secret) IRC channel do not leak out into rooms/channels which ARE on the published room list.
// If a Matrix room is connected to other IRC channels, if ANY single channel is `+s` then remove this Matrix room from the published room list.
// If ALL channels are NOT `+s`, then publish this Matrix room on the published room list.
IrcBridge.prototype._solveVisibility = Promise.coroutine(function*() {
  // ...
}
```